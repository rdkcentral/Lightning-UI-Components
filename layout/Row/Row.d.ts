import lng from '@lightningjs/core';
import { FocusManager } from '../FocusManager/FocusManager';

type transitionObject = {
  delay: number;
  duration: number;
  timingFunction: string;
};

declare class RowStyles {
  itemSpacing?: number;
  scrollIndex?: number;
  alwaysScroll?: boolean;
  neverScroll?: boolean;
  itemTransition?: transitionObject;
}

export declare class Row extends FocusManager {
  itemSpacing?: number;
  scrollIndex?: number;
  alwaysScroll?: boolean;
  neverScroll?: boolean;
  lazyScroll?: boolean;
  autoResizeWidth?: boolean;
  autoResizeHeight?: boolean;
  styles?: RowStyles;

  get itemPosX(): number;
  set itemPosX(x: number);
  get itemPosY(): number;
  set itemPosY(y: number);

  shouldScrollLeft(): boolean;
  shouldScrollRight(): boolean;
  onScreenEffect(): void;
  transitionDone(): void;
  appendItemsAt(items: lng.Component[], index: number): void;
  prependItems(items: lng.Component[]): void;
  removeItemAt(index: number): void;

  $itemChanged(): void;
}
