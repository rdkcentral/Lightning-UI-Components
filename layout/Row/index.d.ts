import lng from '@lightningjs/core';
import FocusManager from '../FocusManager';

type transitionObject = {
  delay: number;
  duration: number;
  timingFunction: string;
};

export interface RowStyles {
  itemSpacing?: number;
  scrollIndex?: number;
  alwaysScroll?: boolean;
  neverScroll?: boolean;
  itemTransition?: transitionObject;
}

export default class Row extends FocusManager {
  itemSpacing?: number;
  scrollIndex?: number;
  alwaysScroll?: boolean;
  neverScroll?: boolean;
  lazyScroll?: boolean;
  autoResizeWidth?: boolean;
  autoResizeHeight?: boolean;
  itemPosX?: number;
  itemPosY?: number;
  style: RowStyles;

  shouldScrollLeft(): boolean;
  shouldScrollRight(): boolean;
  onScreenEffect(): void;
  transitionDone(): void;
  appendItemsAt(items: lng.Component[], index: number): void;
  prependItems(items: lng.Component[]): void;
  removeItemAt(index: number): void;

  $itemChanged(): void;
}
