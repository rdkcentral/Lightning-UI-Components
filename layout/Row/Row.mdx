import { Canvas, Story } from '@storybook/addon-docs';
import { DocsLink, GenericType, ObjectFormat } from '../../.storybook/utils';

# Row

Focus management and smooth scrolling for a horizontally-oriented list of items.

## Usage

`Row` expects a list of `items` which will be laid out side-by-side. By default, the height of `Row`'s Item container is the same as the height of the tallest item, and the width is the width passed in, the width of the wrapping FocusManager if available, or the width of the stage by default. You can change the `w` and `h` attributes to manage `width` and `height`, respectively.

```js
import { Row } from '@lightning/ui';

class Basic extends lng.Component {
  static _template() {
    return {
      Row: {
        type: Row,
        items: [
          // list of items
        ]
      }
    };
  }
}
```

<Canvas>
  <Story id="layout-row--basic" />
</Canvas>

### Scrolling

By default, `Row` will scroll horizontally to capture items out of view (or just outside of the `Row` width).

<Canvas>
  <Story id="layout-row--side-scrolling" />
</Canvas>

The `scrollIndex` property determines the index at which scrolling should begin. By default, that index is 0.

```js
{
  type: Row,
  scrollIndex: 1,
  items: [/* items */]
}
```

## API

### Properties

| name                 | type                                                    | readonly | default   | description                                                                                                                                                                          |
| -------------------- | ------------------------------------------------------- | -------- | --------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
| h                    | number                                                  | false    | 0         | height of the `Row`. Automatically set to tallest item `h` after adding `items`                                                                                                      |
| items                | <DocsLink id="lng.Component">lng.Component[]</DocsLink> | false    | []        | components to list in the row                                                                                                                                                        |
| itemSpacing          | number                                                  | false    | 0         | px between items                                                                                                                                                                     |
| scrollIndex          | number                                                  | false    | 0         | Item index at which scrolling begins, provided the sum of item widths is greater than the width of the `Row`                                                                         |
| selected             | <DocsLink id="lng.Component" />                         | true     | -         | selected item                                                                                                                                                                        |
| selectedIndex        | number                                                  | false    | 0         | index of selected item                                                                                                                                                               |
| w                    | number                                                  | false    | `stage.w` | width of the `Row`                                                                                                                                                                   |
| alwaysScroll         | boolean                                                 | false    |           | determines whether the row will stop scrolling as it nears the right to prevent white space                                                                                          |
| neverScroll          | boolean                                                 | false    | false     | if true, the row will never scroll, unless alwaysScroll is set to true, and if false, the row will apply normal scrolling logic                                                      |
| lazyScroll           | boolean                                                 | false    |           | will only scroll the row if the item is off screen and alwaysScroll and neverScroll are both false                                                                                   |
| onScreenItems        | array                                                   | true     |           | contains a list of items that are currently on screen (within the bounds of the container)                                                                                           |
| autoResize           | boolean                                                 | false    | false     | automatically sets the height and width of the row to be the same as the Items container inside                                                                                      |
| startLazyScrollIndex | number                                                  | false    |           | when `lazyScroll` is `true`, index of item in `items`, and items thereafter, at which lazy scrolling should occur at (alwasy scroll functionality will take place before this index) |
| stopLazyScrollIndex  | number                                                  | false    |           | when `lazyScroll` is `true`, index of item in `items`, and items preceding, at which lazy scrolling should occur (alwasy scroll functionality will take place after this index)      |
| lazyUpCount  | number                                                  | false    |           | number of items to create on screen and new items will be created as user presses right on row. If you have a lot of items in a row and most are off screen, set this to the number of items that are displayed on screen by default       |

### Child Item Properties

The `Row` component will utilize the following properties on each individual element in its `items` array:

| name             | type    | readonly | default | description                                                                                         |
| ---------------- | ------- | -------- | ------- | --------------------------------------------------------------------------------------------------- |
| centerInParent   | boolean | false    | false   | if set to true on an individual element in the `items` array, this will vertically center the child |
| extraItemSpacing | number  | false    | 0       | spacing in px following the item in addition to the `itemsSpacing` value                            |

### Methods

#### onScreenEffect(onScreenItems: lng.Component[]): void

A callback that can be overridden to do something with the items that are currently on screen. This will be called on every new render.

#### appendItems(items: lng.Component[]): void

Adds items to the end of the child list

#### appendItemsAt(items: lng.Component[], index: number): void

Adds items at a specified index of the child list

#### prependItems(items: lng.Component[]): void

Adds items to the start of the child list

##### Arguments

| name  | type                                                    | required | default | description             |
| ----- | ------------------------------------------------------- | -------- | ------- | ----------------------- |
| items | <DocsLink id="lng.Component">lng.Component[]</DocsLink> | false    | []      | list of items to append |

#### render(): void

Called when item focus changes. Render is responsible for the scroll behavior
