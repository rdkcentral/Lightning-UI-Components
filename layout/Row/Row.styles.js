export const base = theme => ({
  itemSpacing: theme.layout.gutterX.xs,
  scrollIndex: 0,
  alwaysScroll: false,
  neverScroll: false,
  itemTransition: {
    timingFunction: theme.animations.functionalStandard,
    delay: theme.animations.functionalStandardDelay,
    duration: theme.animations.functionalStandardDuration
  }
});
