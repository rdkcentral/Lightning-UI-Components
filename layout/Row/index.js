import FocusManager from '../FocusManager';
import { getX, getH, delayForAnimation } from '../../utils';
import withStyles from '../../mixins/withThemeStyles';
import withExtensions from '../../mixins/withExtensions';
import styles from './Row.styles';

class Row extends FocusManager {
  static get __componentName() {
    return 'Row';
  }

  static _template() {
    return {
      ...super._template(),
      direction: 'row'
    };
  }

  static get properties() {
    return [
      ...super.properties,
      'itemSpacing',
      'scrollIndex',
      'alwaysScroll',
      'neverScroll',
      'lazyScroll',
      'lazyUpCount',
      'autoResizeWidth',
      'autoResizeHeight',
      'startLazyScrollIndex',
      'stopLazyScrollIndex'
    ];
  }

  _construct() {
    super._construct();
    this._smooth = false;
  }

  _init() {
    if (!this.w) {
      // if width is undefined or 0, set the Row's width
      if (
        // if the Row is a child item in a FocusManager (like Column)
        this.parent &&
        this.parent.parent &&
        this.parent.parent instanceof FocusManager &&
        this.parent.parent.w
      ) {
        this.w = this.parent.parent.w;
      } else {
        let parent = this.p;
        while (parent && !parent.w) {
          parent = parent.p;
        }
        this.w =
          (parent && parent.w) ||
          this.stage.h / this.stage.getRenderPrecision();
      }
    }

    this.Items.transition('x').on(
      'finish',
      this._transitionListener.bind(this)
    );

    super._init();
  }

  _update() {
    this._updateLayout();
  }

  _getItemSpacing() {
    return this._itemSpacing !== undefined
      ? this._itemSpacing
      : this._componentStyles.itemSpacing;
  }

  _getScrollIndex() {
    return this._scrollIndex !== undefined
      ? this._scrollIndex
      : this._componentStyles.scrollIndex;
  }

  _getAlwaysScroll() {
    return this._alwaysScroll !== undefined
      ? this._alwaysScroll
      : this._componentStyles.alwaysScroll;
  }

  _getNeverScroll() {
    return this._neverScroll !== undefined
      ? this._neverScroll
      : this._componentStyles.neverScroll;
  }

  get _itemTransition() {
    return this._componentStyles.itemTransition;
  }

  _focus() {
    this.items.forEach(item => (item.parentFocus = true));
  }

  _unfocus() {
    this.items.forEach(item => (item.parentFocus = false));
  }

  selectNext() {
    this._smooth = true;
    if (this._lazyItems && this._lazyItems.length) {
      delayForAnimation(() => {
        this.appendItems(this._lazyItems.splice(0, 1));
      });
    }
    return super.selectNext();
  }

  selectPrevious() {
    this._smooth = true;
    return super.selectPrevious();
  }

  // TODO: can be documented in API when lastScrollIndex is made public
  shouldScrollLeft() {
    let shouldScroll = false;

    if (this._lastScrollIndex) {
      shouldScroll = this.selectedIndex < this._lastScrollIndex;
      if (
        this._prevLastScrollIndex !== undefined &&
        this._prevLastScrollIndex !== this._lastScrollIndex
      ) {
        shouldScroll = true;
      }
    } else {
      shouldScroll = this.selectedIndex >= this.scrollIndex;
    }

    return this._itemsX < 0 && shouldScroll;
  }

  // TODO: can be documented in API when lastScrollIndex is made public
  shouldScrollRight() {
    const lastChild = this.Items.childList.last;
    return (
      this.selectedIndex > this.scrollIndex &&
      // end of Items container < end of last item
      Math.abs(this._itemsX - this.w) <
        lastChild.x + this.Items.childList.last.w
    );
  }

  // Since this is a Row, scrolling should be done only when focused item (this.selected) is fully visible horizontally
  // as scrolling is happening along X axis. So, if we have a row that has height greater than screen's height, it still
  // can scroll. Method below the '_isComponentHorizontallyVisible' does not take "full" visibility into an account.
  // At the same time, 'isFullyOnScreen' method on the Base class or utils method checks that element is fully visible
  // both vertically and horizontally.
  // At a later time, we should investigate this further.
  _isOnScreenForScrolling(child) {
    if (!child) return false;
    const itemX = child.core.renderContext.px;
    const rowX = this.core.renderContext.px;
    return itemX >= rowX && itemX + child.w <= rowX + this.w;
  }

  _shouldScroll() {
    if (
      this.lazyScroll &&
      (this.selectedIndex <= this.startLazyScrollIndex ||
        this.selectedIndex >= this.stopLazyScrollIndex)
    ) {
      return true;
    }

    let shouldScroll = this.alwaysScroll || this._selectedPastAdded;
    if (!shouldScroll && !this.neverScroll) {
      const isCompletelyOnScreen = this._isOnScreenForScrolling(this.selected);

      if (this.lazyScroll) {
        shouldScroll = !isCompletelyOnScreen;
      } else {
        const lastChild = this.Items.childList.last;
        shouldScroll =
          lastChild &&
          (this.shouldScrollLeft() ||
            this.shouldScrollRight() ||
            !isCompletelyOnScreen);
      }
    }
    return shouldScroll;
  }

  _getPrependedOffset() {
    this._selectedPastAdded = false;
    return this.Items.x - this._totalAddedWidth;
  }

  _getLazyScrollX(prev) {
    const prevIndex = this.Items.childList.getIndex(this.prevSelected);

    if (this._selectedPastAdded) {
      return this._getPrependedOffset();
    }

    if (this.selectedIndex <= this.startLazyScrollIndex) {
      // if navigating on items before start lazy scroll index, use normal scroll logic
      return this._getScrollX();
    } else if (
      this.selectedIndex >= this.stopLazyScrollIndex &&
      this.selectedIndex < prevIndex
    ) {
      // if navigating left on items after stop lazy scroll index, only shift by size of prev item
      const currItemsX = this.Items.transition('x')
        ? this.Items.transition('x').targetValue
        : this.Items.x;

      return (
        currItemsX +
        (this.prevSelected.w +
          this.itemSpacing +
          (this.selected.extraItemSpacing || 0))
      );
    } else if (prev) {
      // otherwise, no start/stop indexes, perform normal lazy scroll
      let itemsContainerX;
      const prevIndex = this.Items.childList.getIndex(prev);
      if (prevIndex === -1) {
        // No matches found in childList, start set x to 0
        return;
      }
      if (prevIndex > this.selectedIndex) {
        itemsContainerX = -this.selected.x;
      } else if (prevIndex < this.selectedIndex) {
        itemsContainerX = this.w - this.selected.x - this.selected.w;
      }

      return itemsContainerX;
    }
    // if no prev item or start/stop index, default to normal scroll logic
    return this._getScrollX();
  }

  _getScrollX() {
    if (this._selectedPastAdded) {
      return this._getPrependedOffset();
    }
    let itemsContainerX;
    let itemIndex = this.selectedIndex - this.scrollIndex;
    itemIndex = itemIndex < 0 ? 0 : itemIndex;
    if (itemIndex === this._firstFocusableIndex()) {
      itemIndex = 0;
    }

    if (this.Items.children[itemIndex]) {
      itemsContainerX = this.Items.children[itemIndex].transition('x')
        ? -this.Items.children[itemIndex].transition('x').targetValue
        : -this.Items.children[itemIndex].x;
    }

    return itemsContainerX;
  }

  render(next, prev) {
    if (this.plinko && prev && prev.selected) {
      next.selectedIndex = this._getIndexOfItemNear(next, prev);
    }

    this._prevLastScrollIndex = this._lastScrollIndex;

    let itemsContainerX;
    if (!this.Items.children.length) {
      itemsContainerX = this.itemPosX;
    } else if (this._shouldScroll()) {
      itemsContainerX =
        this.lazyScroll && prev
          ? this._getLazyScrollX(prev)
          : this._getScrollX();
    }
    if (itemsContainerX !== undefined) {
      if (this._smooth) {
        this.Items.smooth = {
          x: [itemsContainerX, this._itemTransition]
        };
      } else {
        this.Items.x = itemsContainerX;
        this._updateTransitionTarget(this.Items, 'x', itemsContainerX);
      }
    }

    this.onScreenEffect(this.onScreenItems);
  }

  _updateLayout() {
    let nextX = 0;
    let nextH = 0;
    // layout items in row
    for (let i = 0; i < this.Items.children.length; i++) {
      const child = this.Items.children[i];
      nextH = Math.max(nextH, getH(child));
      if (this._smooth) {
        child.smooth = { x: [nextX, this._itemTransition] };
      } else {
        child.patch({ x: nextX });
        this._updateTransitionTarget(child, 'x', nextX);
      }
      nextX += child.w;
      if (i < this.Items.children.length - 1) {
        const extraItemSpacing = child.extraItemSpacing || 0;
        nextX += this.itemSpacing + extraItemSpacing;
      }

      if (child.centerInParent) {
        // if the child is another focus manager, check the height of the item container
        const childHeight = (child.Items && child.Items.h) || child.h;
        // only center the child if it is within the bounds of this focus manager
        if (childHeight < this.h) {
          child.y = (this.h - childHeight) / 2;
        }
      }
    }

    const itemChanged = this.Items.h !== nextH || this.Items.w !== nextX;
    this.Items.patch({ h: nextH, w: nextX + (this._totalAddedWidth || 0) });

    if (this.autoResizeWidth) {
      this.w = this.Items.w;
    }
    if (this.autoResizeHeight) {
      this.h = this.Items.h;
    }

    const lastChild = this.Items.childList.last;
    const endOfLastChild = lastChild ? getX(lastChild) + lastChild.w : 0;
    const scrollOffset = (this.Items.children[this.scrollIndex] || { x: 0 }).x;

    // determine when to stop scrolling right
    if (this.alwaysScroll) {
      this._lastScrollIndex = this.Items.children.length - 1;
    } else if (endOfLastChild > this.w) {
      for (let i = this.Items.children.length - 1; i >= 0; i--) {
        const child = this.Items.children[i];
        const childX = getX(child);
        if (childX + this.w - scrollOffset > endOfLastChild) {
          this._lastScrollIndex = i;
        } else {
          break;
        }
      }
    }
    itemChanged && this.fireAncestors('$itemChanged');
    this.render(this.selected, this.prevSelected);
  }

  get _itemsX() {
    return getX(this.Items);
  }

  _isOnScreen(child) {
    if (!child) return false;

    return this._isComponentHorizontallyVisible(child);
  }

  appendItems(items = []) {
    const itemHeight = this.renderHeight;
    this._smooth = false;

    if (items.length > this.lazyUpCount + 2) {
      this._lazyItems = items.splice(this.lazyUpCount + 2);
    }

    items.forEach(item => {
      item.parentFocus = this.hasFocus();
      item = this.Items.childList.a(item);
      item.h = item.h || itemHeight;
    });
    this.stage.update();
    this._requestUpdateDebounce();
    this._refocus();
  }

  appendItemsAt(items = [], idx) {
    const addIndex = Number.isInteger(idx) ? idx : this.Items.children.length;
    this._smooth = false;
    this._lastAppendedIdx = addIndex;
    this._totalAddedWidth = 0;

    items.forEach((item, itemIdx) => {
      this.Items.childList.addAt(
        {
          ...item,
          parentFocus: this.hasFocus(),
          h: item.h || this.Items.h
        },
        addIndex + itemIdx
      );
      const extraItemSpacing = item.extraItemSpacing || 0;
      this._totalAddedWidth += item.w + this.itemSpacing + extraItemSpacing;
    });

    if (this.selectedIndex >= this._lastAppendedIdx) {
      this._selectedPastAdded = true;
      this._selectedIndex += items.length;
    }

    this._update();
    this._refocus();
  }

  prependItems(items) {
    this.appendItemsAt(items, 0);
  }

  removeItemAt(index) {
    this._smooth = false;
    this.Items.childList.removeAt(index);

    if (
      this.selectedIndex > index ||
      this.selectedIndex === this.Items.children.length
    ) {
      this._selectedIndex--;
    }

    this._update();
    this._refocus();
  }

  $itemChanged() {
    this._requestUpdateDebounce();
  }

  _transitionListener() {
    this._smooth = false;
    this.transitionDone();
  }

  // can be overridden
  onScreenEffect() {}

  // can be overridden
  transitionDone() {}
}

export default withExtensions(withStyles(Row, styles));
