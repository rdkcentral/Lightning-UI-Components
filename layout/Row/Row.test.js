import lng from '@lightningjs/core';
import TestUtils from '../../test/lightning-test-utils';
import Row from '.';

const baseItem = {
  type: lng.Component,
  w: 80,
  h: 80
};
const items = [
  { ...baseItem },
  { ...baseItem },
  { ...baseItem },
  { ...baseItem },
  { ...baseItem }
];

const createRow = TestUtils.makeCreateComponent(Row, {
  title: 'My Row',
  h: 80,
  w: 400,
  upCount: 5,
  signals: {
    selectedChange: 'selectedChangeMock'
  },
  debounceDelay: 0,
  items
});

describe('Row', () => {
  let testRenderer, row;

  beforeEach(() => {
    [row, testRenderer] = createRow();
    return row._whenEnabled;
  });

  afterEach(() => {
    row = null;
    testRenderer = null;
  });

  it('should render', () => {
    const tree = testRenderer.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render with no items', () => {
    row.items = [];
    testRenderer.update();
    const tree = testRenderer.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should set focus on first item', () => {
    expect(row.items[0].hasFocus()).toBe(true);
  });

  it('should return false on keyLeft', () => {
    expect(row._handleLeft()).toBe(false);
  });

  it('should set focus on next item on keyRight', () => {
    testRenderer.keyPress('Right');
    expect(row.items[1].hasFocus()).toBe(true);
  });

  describe('itemSpacing', () => {
    it('should initialize spacing between items', () => {
      const itemSpacing = 20;
      [row, testRenderer] = createRow({ itemSpacing });
      const item = row.items[1];
      row._update.flush();

      expect(item.x).toBe(row.items[0].w + itemSpacing);
    });

    it('should set spacing', () => {
      const itemSpacing = 100;
      const item = row.items[1];
      row.itemSpacing = itemSpacing;
      row._update.flush();

      const x = item.x;
      expect(x).toBe(row.items[0].w + itemSpacing);
    });
  });

  describe('parentFocus', () => {
    it('should tell the items when row focus changes', () => {
      row._focus();
      expect(row.items[0].parentFocus).toBe(true);
      row._unfocus();
      expect(row.items[0].parentFocus).toBe(false);
    });
  });

  describe('appendItems', () => {
    it('adds items to the item list', () => {
      const ITEMS_LENGTH = row.items.length;
      row.appendItems([{ ...baseItem }, { ...baseItem }]);

      expect(row.items.length).toBe(ITEMS_LENGTH + 2);
    });

    it('items are added outside of the viewable bounds', () => {
      const item = { ...baseItem };
      expect(row.items.length).toBe(5);
      row.appendItems([item]);
      expect(row.items.length).toBe(6);
      expect(row.items[row.items.length - 1].x >= row.x + row.w).toBe(true);
    });

    it('has works with no items', () => {
      const { length } = row.items;
      row.appendItems();
      expect(row.items.length).toBe(length);
    });

    it('defaults item height to row renderHeight', () => {
      const item = { ...baseItem };
      delete item.h;

      row.appendItems([item]);
      expect(row.items[row.items.length - 1].h).toBe(row.h);
    });
  });

  describe('wrapping', () => {
    it('should default to false and should not wrap', () => {
      row._selectedIndex = 4;
      testRenderer.keyPress('Right');
      expect(row._selectedIndex).toBe(4);
    });

    it('should focus last item after Left on first item', () => {
      row.wrapSelected = true;
      testRenderer.keyPress('Left');
      expect(row._selectedIndex).toBe(4);
    });

    it('should focus first item after Right on last item', () => {
      row.wrapSelected = true;
      row._selectedIndex = 4;
      testRenderer.keyPress('Right');
      expect(row._selectedIndex).toBe(0);
    });
  });

  describe('listeners', () => {
    it('should listen for $itemChanged', () => {
      const item1X = row.items[1].x;
      row.scrollTransition = { duration: 0 };
      row.items[0].w += 200;
      row.$itemChanged();
      row._update.flush();
      testRenderer.update();
      expect(row.items[1].x).toBe(item1X + 200);
    });
  });

  describe('scrolling', () => {
    beforeEach(() => {
      row.scrollTransition = { duration: 0 };
      row.items = [...items, ...items];
    });

    it('should scroll long rows', done => {
      expect(row.Items.x).toBe(0);
      testRenderer.keyPress('Right');
      row._whenEnabled.then(() => {
        expect(row._selectedIndex).toBe(1);
        expect(row.Items.transition('x').targetValue).toBe(-row.selected.x);
        done();
      });
    });

    it('should reset the Items x position when there are no items', done => {
      row.itemPosX = 100;
      row.items = [];
      testRenderer.keyPress('Right');
      row._whenEnabled.then(() => {
        expect(row.Items.x).toBe(100);
        done();
      });
    });

    // TODO: Fix - released to get h
    // it('should pass on screen items to onScreenEffect', done => {
    //   row.w = 200;
    //   const onScreenEffect = jest.fn();
    //   row.onScreenEffect = onScreenEffect;
    //   testRenderer.keyPress('Right');
    //   testRenderer.update();

    //   row._whenEnabled.then(() => {
    //     expect(onScreenEffect).toBeCalled();
    //     const onScreenItems = onScreenEffect.mock.calls[0][0].map(item =>
    //       row.items.indexOf(item)
    //     );
    //     const expected = row.items
    //       .filter(item => {
    //         const x1 = item.x;
    //         const x2 = item.x + item.w;
    //         return (
    //           x2 + row.Items.transition('x').targetValue > 0 &&
    //           x1 + row.Items.transition('x').targetValue < row.w
    //         );
    //       })
    //       .map(item => row.items.indexOf(item));
    //     expect(onScreenItems).toEqual(expected);
    //     done();
    //   });
    // });

    describe('with scrollMount=0.5', () => {
      beforeEach(() => {
        row.scrollMount = 0.5;
        TestUtils.fastForward(row.items);
        testRenderer.update();
      });

      describe('navigating right', () => {
        it('does not scroll if selected index < start scroll index', () => {
          const expectedItems = expect.arrayContaining(
            row.items.map(({ x }) => x)
          );
          testRenderer.keyPress('Right');
          TestUtils.fastForward(row.items);
          testRenderer.update();

          expect(row.items.map(({ x }) => x)).toEqual(expectedItems);
        });

        it('shifts items to the left if selected index > start scroll index', done => {
          expect(row.items.map(({ x }) => x)).toEqual(
            expect.arrayContaining([
              0, 80, 160, 240, 320, 400, 480, 560, 640, 720
            ])
          );

          testRenderer.keyPress('Right');

          row._whenEnabled.then(() => {
            testRenderer.update();
            expect(row.selected.x).toBe(80);
            expect(row._itemsX).toBe(-80);
            done();
          });
        });

        it('does not scroll if last item is already in view', () => {
          testRenderer.keyPress('Right');
          testRenderer.keyPress('Right');
          testRenderer.keyPress('Right');
          testRenderer.keyPress('Right');
          testRenderer.keyPress('Right');
          testRenderer.keyPress('Right');
          testRenderer.keyPress('Right');
          TestUtils.fastForward(row.items);
          testRenderer.update();

          const expectedItems = expect.arrayContaining(
            row.items.map(({ x }) => x)
          );
          expect(row.selectedIndex).toBe(7);

          testRenderer.keyPress('Right');
          TestUtils.fastForward(row.items);
          testRenderer.update();

          expect(row.selectedIndex).toBe(8);
          expect(row.items.map(({ x }) => x)).toEqual(expectedItems);
        });
      });
    });

    describe('with neverScroll set to true', () => {
      it('should never scroll the row', done => {
        row.neverScroll = true;
        expect(row.Items.x).toBe(0);
        testRenderer.keyPress('Right');
        testRenderer.keyPress('Right');
        testRenderer.keyPress('Right');
        testRenderer.update();
        row._whenEnabled.then(() => {
          expect(row._selectedIndex).toBe(3);
          expect(row.Items.transition('x').targetValue).toBe(0);
          done();
        });
      });
    });
  });
});
