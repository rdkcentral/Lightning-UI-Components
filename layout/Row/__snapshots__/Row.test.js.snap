// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Row should render 1`] = `
Object {
  "Component": Object {
    "Items": Object {
      "active": false,
      "alpha": 1,
      "attached": true,
      "boundsMargin": null,
      "children": Object {
        "Element-4": Object {
          "active": false,
          "alpha": 1,
          "attached": true,
          "boundsMargin": null,
          "clipping": false,
          "color": 4294967295,
          "enabled": true,
          "flex": false,
          "flexItem": false,
          "h": 80,
          "hasFinalFocus": true,
          "hasFocus": true,
          "isComponent": true,
          "mount": 0,
          "mountX": 0,
          "mountY": 0,
          "pivot": 0.5,
          "pivotX": 0.5,
          "pivotY": 0.5,
          "ref": null,
          "renderOfScreen": undefined,
          "renderToTexture": false,
          "scale": 1,
          "scaleX": 1,
          "scaleY": 1,
          "state": "",
          "tag": [Function],
          "type": "Component",
          "visible": true,
          "w": 80,
          "x": 0,
          "y": 0,
          "zIndex": 0,
        },
        "Element-5": Object {
          "active": false,
          "alpha": 1,
          "attached": true,
          "boundsMargin": null,
          "clipping": false,
          "color": 4294967295,
          "enabled": true,
          "flex": false,
          "flexItem": false,
          "h": 80,
          "hasFinalFocus": false,
          "hasFocus": false,
          "isComponent": true,
          "mount": 0,
          "mountX": 0,
          "mountY": 0,
          "pivot": 0.5,
          "pivotX": 0.5,
          "pivotY": 0.5,
          "ref": null,
          "renderOfScreen": undefined,
          "renderToTexture": false,
          "scale": 1,
          "scaleX": 1,
          "scaleY": 1,
          "state": "",
          "tag": [Function],
          "type": "Component",
          "visible": true,
          "w": 80,
          "x": 104,
          "y": 0,
          "zIndex": 0,
        },
        "Element-6": Object {
          "active": false,
          "alpha": 1,
          "attached": true,
          "boundsMargin": null,
          "clipping": false,
          "color": 4294967295,
          "enabled": true,
          "flex": false,
          "flexItem": false,
          "h": 80,
          "hasFinalFocus": false,
          "hasFocus": false,
          "isComponent": true,
          "mount": 0,
          "mountX": 0,
          "mountY": 0,
          "pivot": 0.5,
          "pivotX": 0.5,
          "pivotY": 0.5,
          "ref": null,
          "renderOfScreen": undefined,
          "renderToTexture": false,
          "scale": 1,
          "scaleX": 1,
          "scaleY": 1,
          "state": "",
          "tag": [Function],
          "type": "Component",
          "visible": true,
          "w": 80,
          "x": 208,
          "y": 0,
          "zIndex": 0,
        },
        "Element-7": Object {
          "active": false,
          "alpha": 1,
          "attached": true,
          "boundsMargin": null,
          "clipping": false,
          "color": 4294967295,
          "enabled": true,
          "flex": false,
          "flexItem": false,
          "h": 80,
          "hasFinalFocus": false,
          "hasFocus": false,
          "isComponent": true,
          "mount": 0,
          "mountX": 0,
          "mountY": 0,
          "pivot": 0.5,
          "pivotX": 0.5,
          "pivotY": 0.5,
          "ref": null,
          "renderOfScreen": undefined,
          "renderToTexture": false,
          "scale": 1,
          "scaleX": 1,
          "scaleY": 1,
          "state": "",
          "tag": [Function],
          "type": "Component",
          "visible": true,
          "w": 80,
          "x": 312,
          "y": 0,
          "zIndex": 0,
        },
        "Element-8": Object {
          "active": false,
          "alpha": 1,
          "attached": true,
          "boundsMargin": null,
          "clipping": false,
          "color": 4294967295,
          "enabled": true,
          "flex": false,
          "flexItem": false,
          "h": 80,
          "hasFinalFocus": false,
          "hasFocus": false,
          "isComponent": true,
          "mount": 0,
          "mountX": 0,
          "mountY": 0,
          "pivot": 0.5,
          "pivotX": 0.5,
          "pivotY": 0.5,
          "ref": null,
          "renderOfScreen": undefined,
          "renderToTexture": false,
          "scale": 1,
          "scaleX": 1,
          "scaleY": 1,
          "state": "",
          "tag": [Function],
          "type": "Component",
          "visible": true,
          "w": 80,
          "x": 416,
          "y": 0,
          "zIndex": 0,
        },
      },
      "clipping": false,
      "color": 4294967295,
      "enabled": true,
      "flex": false,
      "flexItem": false,
      "h": 80,
      "isComponent": undefined,
      "mount": 0,
      "mountX": 0,
      "mountY": 0,
      "pivot": 0.5,
      "pivotX": 0.5,
      "pivotY": 0.5,
      "ref": "Items",
      "renderOfScreen": undefined,
      "renderToTexture": false,
      "scale": 1,
      "scaleX": 1,
      "scaleY": 1,
      "state": undefined,
      "tag": [Function],
      "tags": Array [
        "Items",
      ],
      "visible": true,
      "w": 496,
      "x": 0,
      "y": 0,
      "zIndex": 0,
    },
  },
}
`;

exports[`Row should render with no items 1`] = `
Object {
  "Component": Object {
    "Items": Object {
      "active": true,
      "alpha": 1,
      "attached": true,
      "boundsMargin": null,
      "clipping": false,
      "color": 4294967295,
      "enabled": true,
      "flex": false,
      "flexItem": false,
      "h": 0,
      "isComponent": undefined,
      "mount": 0,
      "mountX": 0,
      "mountY": 0,
      "pivot": 0.5,
      "pivotX": 0.5,
      "pivotY": 0.5,
      "ref": "Items",
      "renderOfScreen": undefined,
      "renderToTexture": false,
      "scale": 1,
      "scaleX": 1,
      "scaleY": 1,
      "state": undefined,
      "tag": [Function],
      "tags": Array [
        "Items",
      ],
      "visible": true,
      "w": 0,
      "x": 0,
      "y": 0,
      "zIndex": 0,
    },
  },
}
`;
